kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseServiceIntegrationTest.java

```diff
@@ -341,40 +311,68 @@ public void testAddDynamicWorkItemTask() {
-            caseClient.addDynamicTask(CONTAINER_ID, "not-existing-case", "ContactCarProducer", "Contact car producer", null);
-            fail("Should have failed because of not existing case case Id.");
-        } catch (KieServicesException e) {
-            // expected
-        }
+        assertClientException(
+                () -> caseClient.addDynamicTask(CONTAINER_ID, BAD_CASE_ID, "ContactCarProducer", "Contact car producer", null),
+                404 , BAD_CASE_ID);
+    }
+
+    @Test
+    public void testAddDynamicTaskToNotExistingStage() throws Exception {
+        String caseId = caseClient.startCase(CONTAINER_ID, CLAIM_CASE_DEF_ID);
+        Assertions.assertThat(caseId).isNotNull();
+        assertClientException(
+                () -> caseClient.addDynamicTaskToStage(CONTAINER_ID, caseId, NON_EXISTENT_STAGE_ID,
+                        "ContactCarProducer", "Contact car producer", null), 404, NON_EXISTENT_STAGE_ID);
+    }
+
+    @Test
+    public void testAddDynamicTaskToStageNotExistingCase() throws Exception {
+        assertClientException(
+                () -> caseClient.addDynamicTaskToStage(CONTAINER_ID, NON_EXISTENT_CASE_ID, "Stage One",
+                        "ContactCarProducer", "Contact car producer", null), 404, NON_EXISTENT_CASE_ID);
+    }
+
+    @Test
+    public void testAddDynamicTaskToStageInvalidContainer() throws Exception {
+        String caseId = caseClient.startCase(CONTAINER_ID, CLAIM_CASE_DEF_ID);
+        Assertions.assertThat(caseId).isNotNull();
+
+        List<CaseStage> caseStages = caseClient.getStages(CONTAINER_ID, caseId, false, 0, 100);
+        Assertions.assertThat(caseStages).isNotNull();
+        Assertions.assertThat(caseStages).isNotEmpty();
+
+        CaseStage firstCaseStage = caseStages.iterator().next();
+        String firstStageId = firstCaseStage.getIdentifier();
+        assertClientException(
+                () -> caseClient.addDynamicTaskToStage(BAD_CONTAINER_ID, caseId, firstStageId,
+                        "ContactCarProducer", "Contact car producer", null), 404, BAD_CONTAINER_ID);
     }
 
     private void assertCarInsuranceCaseInstance(CaseInstance caseInstance, String caseId, String owner) {
-        assertNotNull(caseInstance);
-        assertEquals(caseId, caseInstance.getCaseId());
-        assertEquals(CLAIM_CASE_DEF_ID, caseInstance.getCaseDefinitionId());
-        assertEquals(CLAIM_CASE_DESRIPTION, caseInstance.getCaseDescription());
-        assertEquals(owner, caseInstance.getCaseOwner());
-        assertEquals(ProcessInstance.STATE_ACTIVE, caseInstance.getCaseStatus().intValue());
-        assertNotNull(caseInstance.getStartedAt());
-        assertNull(caseInstance.getCompletedAt());
-        assertEquals("", caseInstance.getCompletionMessage());
-        assertEquals(CONTAINER_ID, caseInstance.getContainerId());
+        Assertions.assertThat(caseInstance).isNotNull();
+        Assertions.assertThat(caseInstance.getCaseId()).isEqualTo(caseId);
```

Comment:
Not intending to hold this PR back, just FYI... Whenever you're using multiple assertions on the same object that do not have dependencies between one another, try `SoftAssertions`:

`SoftAssertions.assertSoftly(softly ->{
    softly.assertThat(caseInstance.getCaseDefinitionId()).isEqualTo(CLAIM_CASE_DEF_ID);
    softly.assertThat(caseInstance.getCaseDescription()).isEqualTo(CLAIM_CASE_DESRIPTION);
    softly.assertThat(caseInstance.getCaseOwner()).isEqualTo(owner);
    softly.assertThat(caseInstance.getCaseStatus().intValue()).isEqualTo(ProcessInstance.STATE_ACTIVE);
    softly.assertThat(caseInstance.getStartedAt()).isNotNull();
    softly.assertThat(caseInstance.getCompletedAt()).isNull();
    softly.assertThat(caseInstance.getCompletionMessage()).isEqualTo("");
    softly.assertThat(caseInstance.getContainerId()).isEqualTo(CONTAINER_ID);
});`

The benefit here is that all the assertions will be evaluated at once, and then reported as such. Instead of evaluating one-by-one and failing when the first assertion fails.